File: incre-tests/fusion/identities/page6.f
Config VerifyBase = 300;

Inductive List = nil Unit | cons {Int, List};
Inductive List2D = nil2D Unit | cons2D {List, List2D};
Inductive List3D = nil3D Unit | cons3D {List2D, List3D};

head = \xs: List2D.
  match xs with
    nil2D _ -> nil unit
  | cons2D {h, t} -> h
  end;

tails = fix (
  \f: List -> Compress List2D. \xs: List.
  match xs with
    nil _ -> cons2D {xs, nil2D unit}
  | cons {h, t} -> cons2D {xs, f t}
  end
);

append = \w: Int. fix (
  \f: List -> List. \xs: List.
  match xs with
    nil _ -> cons {w, nil unit}
  | cons {h, t} -> cons {h, f t}
  end
);

scanl = \f: Int -> List -> List. \init: List. (fix (
  \g: List -> List -> List2D. \now: List. \xs: List.
  match xs with
    nil _ -> cons2D {now, nil2D unit}
  | cons {h, t} -> cons2D {now, g (f h now) t}
  end
)) init;

inits = scanl append (nil unit);

map = \g: List -> List2D. fix (
  \f: List2D -> List3D. \xs: List2D.
  match xs with
    nil2D _ -> nil3D unit
  | cons2D {h, t} -> cons3D {g h, f t}
  end
);

concat = fix (
  \f: List3D -> List2D. \xs: List3D.
  match xs with
    nil3D _ -> nil2D unit
  | cons3D {h, t} ->
    (fix (
      \g: List2D -> List2D. \ys: List2D.
      match ys with
        nil2D _ -> f t
      | cons2D {h1, t1} -> cons2D {h1, g t1}
      end
    )) h
  end
);

segs = \xs: List. concat (map inits (tails xs));

mapL = \g: List -> Int. fix (
  \f: List2D -> List. \xs: List2D.
  match xs with
    nil2D _ -> nil unit
  | cons2D {h, t} -> cons {g h, f t}
  end
);

sum = fix (
  \f: List -> Int. \xs: List.
  match xs with
    nil _ -> 0
  | cons {h, t} -> + h (f t)
  end
);

max = \a: Int. \b: Int. if < a b then b else a;
maximum = fix (
  \f: List -> Int. \xs: List.
  match xs with
    nil _ -> 0
  | cons {h, t} -> max h (f t)
  end
);

maxsum = \xs: List2D. maximum (mapL sum xs);

mss = \xs: List. maxsum (segs xs);

File: ml-benchmark/fusion/identities/page6.ml
type list_ =
  | Nil of unit
  | Cons of { h: int; t: list_ }

type list2d =
  | Nil2D of unit
  | Cons2D of { h: list_; t: list2d }

type list3d =
  | Nil3D of unit
  | Cons3D of { h: list2d; t: list3d }

let head xs =
  match xs with
  | Nil2D _ -> Nil ()
  | Cons2D r -> r.h

let rec tails xs =
  match xs with
  | Nil _ -> Cons2D { h = xs; t = Nil2D () }
  | Cons r -> Cons2D { h = xs; t = tails r.t }

let append w =
  let rec f xs =
    match xs with
    | Nil _ -> Cons { h = w; t = Nil () }
    | Cons r -> Cons { h = r.h; t = f r.t }
  in
  f

let scanl f init =
  let rec g now xs =
    match xs with
    | Nil _ -> Cons2D { h = now; t = Nil2D () }
    | Cons r -> Cons2D { h = now; t = g (f r.h now) r.t }
  in
  g init

let inits = scanl append (Nil ())

let map g =
  let rec f xs =
    match xs with
    | Nil2D _ -> Nil3D ()
    | Cons2D r -> Cons3D { h = g r.h; t = f r.t }
  in
  f

let rec concat xs =
  match xs with
  | Nil3D _ -> Nil2D ()
  | Cons3D r ->
      let rec g ys =
        match ys with
        | Nil2D _ -> concat r.t
        | Cons2D r1 -> Cons2D { h = r1.h; t = g r1.t }
      in
      g r.h

let segs xs = concat (map inits (tails xs))

let mapL g =
  let rec f xs =
    match xs with
    | Nil2D _ -> Nil ()
    | Cons2D r -> Cons { h = g r.h; t = f r.t }
  in
  f

let rec sum xs =
  match xs with
  | Nil _ -> 0
  | Cons r -> r.h + sum r.t

let max a b = if a < b then b else a

let rec maximum xs =
  match xs with
  | Nil _ -> 0
  | Cons r -> max r.h (maximum r.t)

let maxsum xs = maximum (mapL sum xs)

let mss xs = maxsum (segs xs)